### PR TITLE
[LTO] Introduce a helper lambda in gatherImportedSummariesForModule (NFC)

### PR DESCRIPTION
**Pull request by kazutakahirata**
This patch forward ports the heterogeneous std::map::operator[]() from
C++26 so that we can look up the map without allocating an instance of
std::string when the key-value pair exists in the map.

The background is as follows.  I'm planning to reduce the memory
footprint of ThinLTO indexing by changing ImportMapTy, the data
structure used for an import list.  The new list will be a hash set of
tuples (SourceModule, GUID, ImportType) represented in a space
efficient manner.  That means that as we iterate over the hash set, we
encounter SourceModule as many times as GUID.  We don't want to create
a temporary instance of std::string every time we look up
ModuleToSummariesForIndex like:

  auto &SummariesForIndex = ModuleToSummariesForIndex[std::string(ILI.first)];

This patch removes the need to create the temporaries by enabling the
hetegeneous lookup with std::set<K, V, std::less<>> and forward
porting std::map::operator[]() from C++26.

### CODE REVIEW

File: llvm/include/llvm/IR/ModuleSummaryIndex.h [hunk 1]
```diff
@@ -1293,7 +1293,8 @@ using GVSummaryMapTy = DenseMap<GlobalValue::GUID, GlobalValueSummary *>;
 
 /// Map of a module name to the GUIDs and summaries we will import from that
 /// module.
-using ModuleToSummariesForIndexTy = std::map<std::string, GVSummaryMapTy>;
+using ModuleToSummariesForIndexTy =
+    std::map<std::string, GVSummaryMapTy, std::less<>>;
```
Comment (teresajohnson) [inline llvm/include/llvm/IR/ModuleSummaryIndex.h:1297]: Is this change needed, since std::less is the default?

**Review by teresajohnson — APPROVED** [review]
lgtm with a question below

File: llvm/include/llvm/IR/ModuleSummaryIndex.h (hunk 1 above)
Comment (kazutakahirata) [reply llvm/include/llvm/IR/ModuleSummaryIndex.h:1297]: Yes.  The default comparison class for `std::map<Key>` is `std::less<Key>`, which provides:

```
  bool operator()(const Key &L, const Key &R) const;
```

Notice that `L` and `R` are of the same type, namely `const Key &`.

In contrast, `std::less<>` provides:

```
  template <typename T1, typename T2>
  bool operator()(const T1 &L, const T2 &R) const;
```

Here, `L` and `R` can be of two different types, which is exactly what we need to compare `std::string` and `StringRef`.
